miden-lib/asm/sat/prologue.masm

```diff
@@ -0,0 +1,416 @@
+
+    # load expected hash from memory
+    swapw drop drop drop dup movdn.5 push.4 add padw movup.4 mem_loadw
+
+    # assert that the computed hash matches the expected hash
+    assert_eqw
+
+    # compute note hash
+    # ---------------------------------------------------------------------------------------------
+
+    # TODO: This should be optimized using the `hperm` operation
+
+    # serial number hash - serial_hash = hmerge(serial_number, 0)
+    dup push.1 add padw movup.4 mem_loadw padw hmerge
+
+    # hash serial_hash with script hash - merge_script = hmerge(serial_hash, script_hash)
+    dup.4 push.2 add padw movup.4 mem_loadw hmerge
+
+    # hash merge_script with inputs hash - recipient = hmerge(merge_script, inputs_hash)
+    dup.4 push.3 add padw movup.4 mem_loadw hmerge
+
+    # hash recipient with vault hash - note_hash = hmerge(recipient, vault_hash)
+    dup.4 push.4 add padw movup.4 mem_loadw hmerge
+
+    # store note hash in memory and clear stack
+    dup.4 mem_storew dropw drop
+
+    # TODO: assert note hash exists in note db
+end
+
+#! Process the consumed notes data provided via the advice provider. This involves reading the data
+#! from the advice provider and storing it at the appropriate memory addresses. As each note is
+#! consumed its hash and nullifier is computed. The transaction nullifier commitment is computed
+#! via a sequential hash of all (nullifier, script_root) pairs for all consumed notes.
+#!
+#! Stack: []
+#! Advice Tape: [num_cn,
+#!               CN1_SN, CN1_SR, CN1_IR, CN1_VR,
+#!               cn1_na,
+#!               CN1_A1, CN1_A2, ...
+#!
+#!               CN2_SN,CN2_SR, CN2_IR, CN2_VR,
+#!               cn2_na,
+#!               CN2_A1, CN2_A2, ...
+#!               ...]
+#! Output: []
+#!
+#! - num_cn is the number of consumed notes.
+#! - CN1_SN is the serial number of consumed note 1.
+#! - CN1_SR is the script root of consumed note 1.
+#! - CN1_IR is the inputs root of consumed note 1.
+#! - CN1_VR is the vault root of consumed note 1.
+#! - cn1_na is the number of assets in consumed note 1.
+#! - CN1_A1 is the first asset of consumed note 1.
+#! - CN1_A2 is the second asset of consumed note 1.
+proc.process_consumed_notes_data
+    # read the number of consumed notes from the advice provider
+    adv_push.1
+
+    # store the number of consumed notes
+    dup push.CONSUMED_NOTE_NUM_PTR mem_store
+
+    # loop over consumed notes and read data
+    # ---------------------------------------------------------------------------------------------
+
+    # initialize counter for consumed notes
+    push.0
+
+    # check if the number of consumed notes is greater then 0. Conditional for the while loop.
+    dup.1 dup.1 neq
+
+    # loop and read note data from the advice provider
+    while.true
+        dup exec.process_consumed_note
+
+        # increment consumed note counter and check if we should loop again
+        push.1 add dup.1 dup.1 neq
+    end
+
+    # drop counter
+    drop
```

Comment:
Should we explicitly check that the number of consumed notes does not exceed w/e limit we impose?

Reply:
assertion added with limit of 1023